rollup-sequencer/src/index.ts

```diff
@@ -1,151 +1,73 @@
 import "@mangata-finance/types";
-import { Keyring } from "@polkadot/api";
-import { hexToU8a } from "@polkadot/util";
+import type { HeaderExtended } from "@polkadot/api-derive/type/types";
 import "dotenv/config";
-import { createPublicClient, encodeAbiParameters, webSocket } from "viem";
 import { keccak256 } from "viem";
-import rolldownAbi from "./RollDown.json" assert { type: "json" };
-
-type ContractAddress = `0x${string}`;
-
-const mangataContractAddress = process.env
-	.MANGATA_CONTRACT_ADDRESS! as ContractAddress;
-
-const blockNumberDelay = process.env.BLOCK_NUMBER_DELAY! || 5;
-
-function sleep_ms(ms: number) {
-	return new Promise((resolve) => setTimeout(resolve, ms));
-}
+import { MANGATA_NODE_URL, MNEMONIC } from "./common/constants.js";
+import {
+	getApi,
+	getCollator, getSelectedSequencerWithRights,
+	initReadContractWithRetry,
+	print,
+	processDataForL2Update,
+	processPendingRequestsEvents,
+} from "./utils/index.js";
+import { getPublicClient } from "./viem/client.js";
+import { webSocketTransport } from "./viem/transport.js";
 
 async function main() {
 	let lastSubmitted = "";
-	const abi = rolldownAbi.abi;
-	const publicClient = createPublicClient({
-		transport: webSocket(process.env.ETH_CHAIN_URL, {
-			retryCount: 5,
-		}),
-	});
-
-	while (true) {
-		try {
-			const data = (await publicClient.readContract({
-				address: mangataContractAddress,
-				abi: abi,
-				functionName: "getUpdateForL2",
-			})) as any;
-			console.log(data);
-			break;
-		} catch (e) {
-			console.log(`${mangataContractAddress} contract not found`);
-			await sleep_ms(1000);
-		}
-	}
-
-	const api = await Mangata.instance([process.env.MANGATA_NODE_URL!]).api();
-	await api.isReady;
-
-	const keyring = new Keyring({ type: "ethereum" });
-	const collator = keyring.addFromSeed(hexToU8a(process.env.MNEMONIC!));
-
-	await api.derive.chain.subscribeNewHeads(async (header) => {
-		const apiAt = await api.at(header.hash);
-		console.log(`block #${header.number} was authored by ${header.author}`);
-
-		try {
-			const latestBlockNumber = await publicClient.getBlockNumber();
-			const delayedBlockNumber = latestBlockNumber - BigInt(blockNumberDelay);
 
-			console.log(`Latest Block Number: ${latestBlockNumber.toString()}`);
-			console.log(`Delayed Block Number:  ${delayedBlockNumber.toString()}`);
+	const publicClient = getPublicClient({ transport: webSocketTransport });
 
-			const data = (await publicClient.readContract({
-				address: mangataContractAddress,
-				abi: abi,
-				functionName: "getUpdateForL2",
-				blockNumber: delayedBlockNumber,
-			})) as any;
+	const api = await getApi(MANGATA_NODE_URL);
 
-			console.log(util.inspect(data, { depth: null }));
+	await initReadContractWithRetry(publicClient);
 
-			// @ts-ignore
-			const encodedData = encodeAbiParameters(
-				abi.find((e: any) => e!.name === "getUpdateForL2")!.outputs!,
-				[data],
-			);
+	await api.derive.chain.subscribeNewHeads(async (header: HeaderExtended) => {
```

Comment:
we still want to be able to select trigger method either relay chain or eigen layer